pkg/logs/pipeline/provider.go

```diff
@@ -10,59 +10,65 @@ import (
 
 	"github.com/DataDog/datadog-agent/pkg/logs/config"
 	"github.com/DataDog/datadog-agent/pkg/logs/message"
-	"github.com/DataDog/datadog-agent/pkg/logs/processor"
+	"github.com/DataDog/datadog-agent/pkg/logs/restart"
 	"github.com/DataDog/datadog-agent/pkg/logs/sender"
 )
 
 // Provider provides message channels
 type Provider interface {
-	Start(cm *sender.ConnectionManager, auditorChan chan message.Message)
+	Start()
+	Stop()
 	NextPipelineChan() chan message.Message
 }
 
 // provider implements providing logic
 type provider struct {
-	numberOfPipelines int32
-	chanSizes         int
-	pipelinesChans    [](chan message.Message)
-
-	currentChanIdx int32
+	connManager          *sender.ConnectionManager
+	outputChan           chan message.Message
+	numberOfPipelines    int
+	chanSize             int
+	pipelines            []*Pipeline
+	currentPipelineIndex int32
 }
 
 // NewProvider returns a new Provider
-func NewProvider() Provider {
+func NewProvider(connManager *sender.ConnectionManager, outputChan chan message.Message) Provider {
 	return &provider{
+		connManager:       connManager,
+		outputChan:        outputChan,
 		numberOfPipelines: config.NumberOfPipelines,
-		chanSizes:         config.ChanSizes,
-		pipelinesChans:    [](chan message.Message){},
-		currentChanIdx:    0,
+		chanSize:          config.ChanSize,
+		pipelines:         []*Pipeline{},
 	}
 }
 
 // Start initializes the pipelines
-func (p *provider) Start(cm *sender.ConnectionManager, auditorChan chan message.Message) {
-
-	for i := int32(0); i < p.numberOfPipelines; i++ {
-
-		senderChan := make(chan message.Message, p.chanSizes)
-		f := sender.New(senderChan, auditorChan, cm)
-		f.Start()
-
-		processorChan := make(chan message.Message, p.chanSizes)
-		pr := processor.New(
-			processorChan,
-			senderChan,
-			config.LogsAgent.GetString("api_key"),
-			config.LogsAgent.GetString("logset"),
-		)
-		pr.Start()
+func (p *provider) Start() {
+	for i := 0; i < p.numberOfPipelines; i++ {
+		pipeline := NewPipeline(p.chanSize, p.connManager, p.outputChan)
+		pipeline.Start()
+		p.pipelines = append(p.pipelines, pipeline)
+	}
+}
 
-		p.pipelinesChans = append(p.pipelinesChans, processorChan)
+// Stop stops all pipelines in parallel,
+// this call blocks until all pipelines are stopped
+func (p *provider) Stop() {
+	stopper := restart.NewParallelGroup()
+	for _, pipeline := range p.pipelines {
+		stopper.Add(pipeline)
 	}
+	stopper.Stop()
+	p.pipelines = p.pipelines[:0]
 }
 
-// NextPipelineChan returns the next pipeline
+// NextPipelineChan returns the next pipeline input channel
 func (p *provider) NextPipelineChan() chan message.Message {
-	idx := atomic.AddInt32(&p.currentChanIdx, 1)
-	return p.pipelinesChans[idx%p.numberOfPipelines]
+	pipelinesLen := len(p.pipelines)
+	if pipelinesLen == 0 {
+		return nil
+	}
+	index := atomic.AddInt32(&p.currentPipelineIndex, 1)
```

Comment:
how does this behave when we reach the max value for an int 32?

Reply:
I guess we start all over again, I'll change this.